test_compare/cmp.py

```diff
@@ -0,0 +1,679 @@
+            f"{row.style}{row.status_new:^9}{ESCAPE_RESET}{row.separator}"
+        )
+
+
+def compare_level(data_old, data_new, args, depth=0, path=None):
+    only_old, only_new = remove_non_common(data_old, data_new)
+    only_old = [item for item in only_old if filter((path or []) + [item], args)]
+    only_new = [item for item in only_new if filter((path or []) + [item], args)]
+    output = {
+        "time_old": 0,
+        "time_new": 0,
+        "children": [],
+        "only_old": only_old,
+        "only_new": only_new,
+    }
+    if depth == 3:
+        cases = compare_cases(data_old["cases"], data_new["cases"])
+        cases = [case for case in cases if filter(path + [case["name"]], args)]
+        output["children"] = cases
+        output["time_old"] = sum(
+            case["time_old"] for case in cases if case["status_old"] == OK and case["status_new"] == OK
+        )
+        output["time_new"] = sum(
+            case["time_new"] for case in cases if case["status_old"] == OK and case["status_new"] == OK
+        )
+        output["single_case"] = (
+            len(cases) == 1
+            and cases[0]["name"]
+            in [
+                "",
+                path[-1],
+            ]
+            and cases[0]["status_old"] == OK
+            and cases[0]["status_new"] == OK
+        )
+
+        return output
+    for child, child_data in data_old.items():
+        new_path = (path or []) + [child]
+        if not filter(new_path, args):
+            continue
+        child_data_new = data_new[child]
+        child_output = compare_level(
+            child_data,
+            child_data_new,
+            args,
+            depth + 1,
+            new_path,
+        )
+        if child_output["children"]:
+            child_output["name"] = child
+            output["children"].append(child_output)
+            output["time_old"] += child_output["time_old"]
+            output["time_new"] += child_output["time_new"]
+    return output
+
+
+def find_missing(results, level=0):
+    if level == 3:
+        return [
+            {"name": case["name"], "status_old": case["status_old"], "status_new": case["status_new"]}
+            for case in results["children"]
+            if case["status_old"] != case["status_new"]
+        ]
+    only_old = [
+        {"name": result, "status_old": PRESENT, "status_new": MISSING, "children": []} for result in results["only_old"]
+    ]
+    only_new = [
+        {"name": result, "status_old": MISSING, "status_new": PRESENT, "children": []} for result in results["only_new"]
+    ]
+    with_children = [
+        {"name": result["name"], "status_old": PRESENT, "status_new": PRESENT, "children": children}
+        for result, children in ((result, find_missing(result, level + 1)) for result in results["children"])
+        if children
+    ]
+    return only_old + only_new + with_children
+
+
+def find_fails(results, args, unfiltered=False, level=0, path=None):
+    if level == 3:
```

Comment:
hard-coding max recursion level in multiple places, would be better to just always check if we have sub-results or not